File: python/puzzles/y2022/day14.py
import utils
from utils import SparseMatrix, Coord

example = """498,4 -> 498,6 -> 496,6
503,4 -> 502,4 -> 502,9 -> 494,9"""

FALLING = 0
RESTING = 1
DESTROYED = 2
WALL = "█"
SAND = "░"


def draw_line(start: Coord, end: Coord, grid: SparseMatrix):
    start_x, start_y = start
    end_x, end_y = end
    dy = end_y - start_y
    start_x, end_x = min(start_x, end_x), max(start_x, end_x)
    start_y, end_y = min(start_y, end_y), max(start_y, end_y)

    if dy == 0:
        for x in range(start_x, end_x + 1):
            grid[(x, start_y)] = WALL
    else:
        for y in range(start_y, end_y + 1):
            grid[(start_x, y)] = WALL


def parse_input(raw: str) -> SparseMatrix:
    instructions = raw.splitlines()
    grid = SparseMatrix()
    for ins in instructions:
        points = [tuple(map(int, pair.split(","))) for pair in ins.split(" -> ")]
        for ii in range(len(points) - 1):
            start = points[ii]
            end = points[ii + 1]
            draw_line(start, end, grid)
    return grid


def sand_step(pos: Coord, grid: SparseMatrix, floor: int, solid_floor: bool) -> tuple[Coord, int]:
    x, y = pos
    options = [
        (x, y + 1),  # down
        (x - 1, y + 1),  # down_left
        (x + 1, y + 1),  # down_right
    ]
    for x, y in options:
        if (x, y) not in grid:
            if y == floor:
                if solid_floor:
                    return pos, RESTING
                else:
                    return pos, DESTROYED
            else:
                return (x, y), FALLING

    # all movement options blocked; sand came to rest on obstacle
    return pos, RESTING


def sand_trace(origin: Coord, grid: SparseMatrix, floor: int, solid_floor: bool) -> int:
    pos = origin

    status = FALLING
    while status == FALLING:
        new_pos, status = sand_step(pos, grid, floor, solid_floor)
        pos = new_pos

    if status == RESTING:
        grid[pos] = SAND
    return status


@utils.profile
def part1(raw: str):
    grid = parse_input(raw)
    origin = (500, 0)
    floor = max(y for x, y in grid)
    ii = 0
    while True:
        status = sand_trace(origin, grid, floor=floor, solid_floor=False)
        if status == DESTROYED:
            break
        ii += 1
    return ii


@utils.profile
def part2(raw: str):
    # todo: speed this up by ignoring any sand particles that get further left/right than the l/r
    #  bounds of the platforms.
    grid = parse_input(raw)
    origin = (500, 0)
    floor = 2 + max(y for x, y in grid)
    ii = 0
    while True:
        sand_trace(origin, grid, floor=floor, solid_floor=True)
        ii += 1
        if origin in grid:
            break
    return ii


if __name__ == "__main__":
    raw = utils.load_puzzle_input("2022/day14")
    assert part1(raw) == 755
    assert part2(raw) == 29805

File: go/2022/day14.go
package _2022

import (
	. "advent/data_structures/sparse_matrix"
	. "advent/data_structures/coord"

	"advent/utils"
	"fmt"
	"strings"
)

const (
	FALLING   = 0
	RESTING   = 1
	DESTROYED = 2
	WALL      = '█'
	SAND      = '░'
)

func drawLine(start, end Coord, grid SparseMatrix) {
	startX, startY := start[0], start[1]
	endX, endY := end[0], end[1]
	dy := endY - startY
	if startX > endX {
		startX, endX = endX, startX
	}
	if startY > endY {
		startY, endY = endY, startY
	}
	if dy == 0 {
		for x := startX; x < endX+1; x++ {
			coord := Coord{x, startY}
			grid[coord] = WALL
		}
	} else {
		for y := startY; y < endY+1; y++ {
			coord := Coord{startX, y}
			grid[coord] = WALL
		}
	}
}

func parseDay14Input(input string) SparseMatrix {
	instructions := strings.Split(input, "\n")
	grid := SparseMatrix{}
	for _, ins := range instructions {
		points := []Coord{}
		for _, pointStr := range strings.Split(ins, " -> ") {
			xy := strings.Split(pointStr, ",")
			x := utils.ParseInt(xy[0])
			y := utils.ParseInt(xy[1])
			points = append(points, Coord{x, y})
		}
		for ii := 0; ii < len(points)-1; ii++ {
			start := points[ii]
			end := points[ii+1]
			drawLine(start, end, grid)
		}
	}
	return grid
}

func sandStep(
	pos Coord,
	grid SparseMatrix,
	floor int, // y-height of the floor
	solidFloor bool, // can the sand come to rest on the floor
) (newPos Coord, status int) {
	x, y := pos[0], pos[1]
	options := []Coord{
		{x, y + 1},
		{x - 1, y + 1},
		{x + 1, y + 1},
	}
	for _, coord := range options {
		y := coord[1]
		_, occupied := grid[coord]
		if !occupied {
			if y == floor {
				if solidFloor {
					return pos, RESTING
				} else {
					return pos, DESTROYED
				}
			} else {
				return coord, FALLING
			}
		}
	}
	return pos, RESTING
}

func sandTrace(origin Coord, grid SparseMatrix, floor int, solidFloor bool) int {
	// return True if sand came to rest OK; false if it fell off the map
	pos := origin
	status := FALLING
	for status == FALLING {
		pos, status = sandStep(pos, grid, floor, solidFloor)
	}
	if status == RESTING {
		grid[pos] = SAND
	}
	return status
}

func Day14Part1(input string) string {
	grid := parseDay14Input(input)
	origin := Coord{500, 0}
	_, abyss := grid.Ylim()
	ii := 0
	for {
		status := sandTrace(origin, grid, abyss, false)
		if status == DESTROYED {
			break
		}
		ii++
	}
	return fmt.Sprint(ii)
}

func Day14Part2(input string) string {
	grid := parseDay14Input(input)
	origin := Coord{500, 0}
	_, maxY := grid.Ylim()
	floor := 2 + maxY
	ii := 0
	for {
		status := sandTrace(origin, grid, floor, true)
		if status == DESTROYED {
			break
		}
		ii++
		_, ok := grid[origin]
		if ok {
			break
		}
	}
	return fmt.Sprint(ii)
}
